src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -67,13 +75,34 @@ public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationT
                 {
                     while (true)
                     {
-                        if (_state == ParsingState.Ending)
+                        if (_connection == null)
                         {
-                            await ConsumeTrailingHeaders().ConfigureAwait(false);
-                            Finish();
+                            // Fully consumed the response in ReadChunksFromConnectionBuffer.
                             return 0;
                         }
 
+                        if (_state == ParsingState.ExpectChunkData &&
+                            destination.Length >= _connection.ReadBufferSize &&
+                            _chunkBytesRemaining >= (ulong)_connection.ReadBufferSize)
+                        {
+                            // As an optimization, we skip going through the connection's read buffer if both
+                            // the remaining chunk data and the destination buffer are both at least as large
+                            // as the connection buffer.  That avoids an unnecessary copy while still reading
+                            // the maximum amount we'd otherwise read at a time.
+                            Debug.Assert(_connection.RemainingBuffer.Length == 0);
```

Comment:
Nit: this is true even if the if test above fails, right?  Should we move outside the if?

Reply:
No, it's only guaranteed to be true if _state == ParsingState.ExpectChunkData (the subsequent length checks don't matter, of course).

Reply:
Yeah, makes sense.  Thanks.